### PR TITLE
remove apiversion parameter from client constructor

### PR DESCRIPTION
**Pull request by chunyu3**
# Description
Fix https://github.com/Azure/autorest.csharp/issues/2888
Fix https://github.com/Azure/autorest.csharp/issues/2800

- remove apiVersion parameter from client constructor, put it in ClientOptions
- support InputConstant for enum type
- fix issue: exception occur when mix use of template & non  template operation

# Checklist

To ensure a quick review and merge, please ensure:
- [ ] The PR has a understandable title and description explaining the _why_ and _what_.
- [ ] The PR is opened in draft if not ready for review yet.
   - If opened in draft, please allocate sufficient time (24 hours) after moving out of draft for review
- [ ] The branch is recent enough to not have merge conflicts upon creation.

# Ready to Land?
- [ ] Build is completely green
   - Submissions with test failures require tracking issue and approval of a CODEOWNER
- [ ] At least one +1 review by a CODEOWNER
- [ ] All -1 reviews are confirmed resolved by the reviewer 
   - Override/Marking reviews stale must be discussed with CODEOWNERS first

### CODE REVIEW

File: src/AutoRest.CSharp/Common/Input/CadlInputConstantConverter.cs [hunk 1]
```diff
@@ -93,6 +93,22 @@ public static object ReadConstantValue(ref Utf8JsonReader reader, string propert
 
                     }
                     break;
+                case InputEnumType enumType:
+                    switch (enumType.EnumValueType.Kind)
+                    {
+                        case InputTypeKind.String:
+                            value = reader.GetString() ?? throw new JsonException();
```
Comment (lirenhe) [inline src/AutoRest.CSharp/Common/Input/CadlInputConstantConverter.cs:100]: Please include the error message for easy trouble shooting.

File: src/AutoRest.CSharp/Properties/launchSettings.json [hunk 2]
```diff
@@ -420,6 +420,10 @@
       "commandName": "Project",
       "commandLineArgs": "--standalone $(SolutionDir)\\test\\TestProjects\\MgmtSubscriptionNameParameter\\Generated"
     },
+    "MixAPIVersion-Cadl": {
```
Comment (archerzz) [inline src/AutoRest.CSharp/Properties/launchSettings.json:423]: Can we test in the existing test project like [APIVersoin](https://github.com/Azure/autorest.csharp/tree/feature/v3/test/TestProjects/ApiVersion)

File: src/AutoRest.CSharp/Common/Input/CadlInputConstantConverter.cs [hunk 3]
```diff
@@ -93,6 +93,22 @@ public static object ReadConstantValue(ref Utf8JsonReader reader, string propert
...
+                            break;
+                        case InputTypeKind.Int32:
+                            value = reader.GetInt32();
+                            break;
```
Comment (lirenhe) [inline src/AutoRest.CSharp/Common/Input/CadlInputConstantConverter.cs:104]: Just curious, this is not related to the APIversion, right?

File: test/TestProjects/MixAPIVersion-Cadl/Generated/MixApiVersionClient.cs [hunk 4]
```diff
@@ -0,0 +1,338 @@
...
+
+        internal HttpMessage CreateCreateRequest(RequestContent content, RequestContext context)
+        {
+            var message = _pipeline.CreateMessage(context, ResponseClassifier200);
+            var request = message.Request;
+            request.Method = RequestMethod.Post;
+            var uri = new RawRequestUriBuilder();
+            uri.Reset(_endpoint);
+            uri.AppendPath("/pets", false);
+            uri.AppendQuery("apiVersion", "2022-07-09", true);
```
Comment (archerzz) [inline test/TestProjects/MixAPIVersion-Cadl/Generated/MixApiVersionClient.cs:296]: But this `apiVersion` parameter defined in cadl won't be exposed to external?

Comment (chunyu3) [reply test/TestProjects/MixAPIVersion-Cadl/Generated/MixApiVersionClient.cs:296]: It is a special case, the operation is bind to a special apiversion, so it will define in the operation as @query apiVersion: "2022-07-09". So it is not the same as the one in clientOption, we must pass the exact value in Create***Request method.

File: src/AutoRest.CSharp/Properties/launchSettings.json (hunk 2 above)
Comment (chunyu3) [reply src/AutoRest.CSharp/Properties/launchSettings.json:423]: No. I want to define an special case for this mix scenario, And ApiVersion-Cadl is to test an apiversion as path paramter.

File: src/AutoRest.CSharp/Common/Input/CadlInputConstantConverter.cs (hunk 3 above)
Comment (chunyu3) [reply src/AutoRest.CSharp/Common/Input/CadlInputConstantConverter.cs:104]: Yes, It is not for apiVersion. this is to support  InputConstant for enum type.

File: src/AutoRest.CSharp/Common/Input/CadlInputConstantConverter.cs (hunk 1 above)
Comment (chunyu3) [reply src/AutoRest.CSharp/Common/Input/CadlInputConstantConverter.cs:100]: The occurs in json reader and we cannot know the detail, so we just return default exception.

File: src/CADL.Extension/Emitter.Csharp/src/emitter.ts [hunk 5]
```diff
@@ -396,33 +401,21 @@ function createModel(
...
-                            apiVersionInOperation.DefaultValue?.Value
-                        )
-                    ) {
-                        apiVersions.add(
-                            apiVersionInOperation.DefaultValue.Value
-                        );
-                    } else {
-                        if (
-                            apiVersionInOperation.Location ===
+                    if ((!apiVersionInOperation.DefaultValue?.Value ||
```
Comment (lirenhe) [inline src/CADL.Extension/Emitter.Csharp/src/emitter.ts:404]: This logic is complex, could you add some comments?

File: src/AutoRest.CSharp/Common/Input/CadlInputConstantConverter.cs (hunk 1 above)
Comment (lirenhe) [reply src/AutoRest.CSharp/Common/Input/CadlInputConstantConverter.cs:100]: Can we just use the message like:  "The value is null for ......, expect 'string'."

File: src/CADL.Extension/Emitter.Csharp/src/emitter.ts (hunk 5 above)
Comment (chunyu3) [reply src/CADL.Extension/Emitter.Csharp/src/emitter.ts:411]: added

**Review by archerzz — APPROVED** [review]

**Review by lirenhe — APPROVED** [review]
I am still no sure we should do 'throw new JsonException()', but i won't block the merge of this PR.

File: src/AutoRest.CSharp/Common/Input/CadlInputConstantConverter.cs (hunk 1 above)
Comment (chunyu3) [reply src/AutoRest.CSharp/Common/Input/CadlInputConstantConverter.cs:100]: It may not because of the value is null or something. The message is also meaningless. Now we just use the default exception as we did before.
